executable-spec/src/Cardano/Ledger/Spec/STS/Update/Ideation.hs

```diff
@@ -156,7 +161,7 @@ instance HashAlgorithm hashAlgo => STS (IDEATION hashAlgo) where
           sip ∈ range subsips ?! NoSIPToReveal sip
           -- TODO: Revealed SIP must belong to stable submitted SIPs
 
-          Data.sipHash sip ∉ dom wrsips ?! SIPAlreadyRevealed sip
+          Data.sipHash sip ∉ ((dom wrsips) `Set.union` (dom asips))?! SIPAlreadyRevealed sip
```

Comment:
I think we should stop emptying `wrsips`. See also my answer in the previous comment.

Reply:
OK here is my proposal for this issue:

1. We stop emptying `subsips` and make it the "history" of all submitted commit-sips. Moreover, we can change it from
`subsips :: !(Map (Commit hashAlgo) (SIP hashAlgo))`
to `subsips :: !(Map (Commit hashAlgo) (SIPCommit hashAlgo))`, so that we dont keep the actual SIP in there.

2. We make `sipdb :: !(Map (Data.SIPHash hashAlgo) (SIP hashAlgo))`, the "history" of all revealed SIPs.
3. We know that 
```
sipdb `isSubsetOf` subsips
```
4. We leave the `wssips`, `wrsips` and `asips` as is, to combine sip with `slot`, signifying when the event happened. We empty these structures at regular intervals based on `slot`.

So the above checks now become:
a) A committed SIP has not appeared in the past again (it is new)
`(Data.commit sipc) ∉ dom subsips ?! SIPAlreadySubmitted sip`

b) A revealed SIP has not been revealed again in the past (it is new)
`sip ∉ range sipdb ?! SIPAlreadyRevealed sip`

c) A generated SIP has not been generated in the past (it is new - not submitted and nor revealed)
```
-- generate a new (never seen before) sip
sip <- Gen.filter
   (\s -> Data.calcCommit s 
             `Set.notMember` 
              (dom subsips)
   )
   sipGen
```
What do you think?